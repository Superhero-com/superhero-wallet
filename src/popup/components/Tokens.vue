<template>
  <span
    class="tokens"
    :class="[iconSize, { vertical, bright }]"
  >
    <span class="icons">
      <img
        v-if="toToken && !noIcons"
        :src="toToken.img"
        :class="['to-token', { border: toToken.imgBorder }]"
        :title="toToken.symbol"
      >
      <img
        v-if="!noIcons"
        :src="fromToken.img"
        :class="{
          border: fromToken.imgBorder,
          pair: !!toToken,
        }"
        :title="fromToken.symbol"
      >
    </span>
    <span class="symbols">
      <span
        v-if="fromToken"
        class="symbol"
      >
        {{ shrinkString(fromToken.symbol) }}
      </span>
      <span
        v-if="fromToken && toToken"
        class="separator"
      >
        /
      </span>
      <span
        v-if="toToken"
        class="symbol"
      >
        {{ shrinkString(toToken.symbol) }}
      </span>
    </span>
  </span>
</template>

<script>
import { computed, defineComponent } from '@vue/composition-api';
import AeIcon from '../../icons/tokens/ae.svg';
import {
  AETERNITY_COIN_NAME,
  AETERNITY_COIN_SYMBOL,
  AETERNITY_CONTRACT_ID,
  AETERNITY_SYMBOL,
  shrinkString as shrinkStringFactory,
} from '../utils';

const SIZES = ['rg', 'md', 'lg', 'xl'];

export default defineComponent({
  props: {
    /**
     * transactionTokenInfoResolvers []
     */
    tokens: { type: Array, required: true },
    symbolLength: { type: Number, default: 11 },
    doubleSymbolLength: { type: Number, default: 5 },
    iconSize: {
      type: String,
      default: 'rg',
      validator: (val) => SIZES.includes(val),
    },
    vertical: Boolean,
    noIcons: Boolean,
    fullAeSymbol: Boolean,
    bright: Boolean,
  },
  setup(props) {
    function getAvailableCharLength() {
      if (props.tokens?.length < 2) return props.symbolLength;
      const shorterNameLength = [props.tokens[0].symbol.length, props.tokens[1].symbol.length]
        .find((length) => length < props.doubleSymbolLength);
      return shorterNameLength ? props.symbolLength - shorterNameLength : props.doubleSymbolLength;
    }

    function shrinkString(text) {
      const maxLength = getAvailableCharLength();
      return shrinkStringFactory(text, maxLength);
    }

    function mapToken(token) {
      let img = `https://avatars.z52da5wt.xyz/${token.contractId}`;
      let imgBorder = true;

      let { symbol } = token;
      let name = token.symbol;

      if (token.isAe || token.contractId === AETERNITY_CONTRACT_ID) {
        img = AeIcon;
        imgBorder = false;
        symbol = props.fullAeSymbol ? AETERNITY_COIN_SYMBOL : AETERNITY_SYMBOL;
        name = AETERNITY_COIN_NAME;
      }

      return {
        ...token,
        symbol,
        name,
        img,
        imgBorder,
      };
    }

    const fromToken = computed(() => (props.tokens?.[0] ? mapToken(props.tokens[0]) : null));
    const toToken = computed(() => (props.tokens?.[1] ? mapToken(props.tokens[1]) : null));

    return {
      fromToken,
      toToken,
      shrinkString,
    };
  },
});
</script>

<style lang="scss" scoped>
@use '../../styles/variables';
@use '../../styles/typography';

.tokens {
  --icon-size: 16px;

  @extend %face-sans-15-medium;

  color: rgba(variables.$color-white, 0.75);

  &.bright {
    color: variables.$color-white;
  }

  &,
  .symbols,
  .icons {
    display: inline-flex;
    align-items: center;
    align-self: center;
  }

  .symbol {
    @extend %face-sans-16-medium;

    vertical-align: middle;
    white-space: nowrap;
    line-height: 20px;
    letter-spacing: -0.02em;
  }

  .separator {
    margin: 0 1px;
    vertical-align: middle;
  }

  &.md {
    --icon-size: 18px;
  }

  &.lg {
    --icon-size: 24px;
  }

  &.xl {
    --icon-size: 30px;
  }

  img {
    width: var(--icon-size);
    height: var(--icon-size);
    border-radius: calc(var(--icon-size) / 2);
    vertical-align: middle;
    margin-right: 4px;

    &.to-token {
      margin-left: 10px;
    }

    &.pair {
      margin-right: 16px;
      margin-left: -30px;
    }

    &.border {
      border: 0.25px solid rgba(variables.$color-white, 0.75);
    }
  }

  &.vertical {
    flex-direction: column;

    .symbols {
      @extend %face-sans-18-medium;

      .separator {
        color: rgba(variables.$color-white, 0.75);
      }
    }

    .icons {
      margin-bottom: 8px;

      img {
        width: 44px;
        height: 44px;
        margin: 0;
        border-radius: 22px;

        &.pair {
          margin-left: -80px;
        }
      }
    }
  }
}
</style>
