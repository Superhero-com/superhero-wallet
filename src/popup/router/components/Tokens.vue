<template>
  <span class="tokens">
    <img
      :src="fromToken.img"
      :class="{ border: fromToken.imgBorder }"
    >
    <img
      v-if="toToken"
      :src="toToken.img"
      :class="{ border: toToken.imgBorder }"
    >
    <Tooltip
      v-if="fromToken"
      class="symbol"
      :tooltip="fromToken.name"
    >
      {{ shrinkString(fromToken.symbol) }}
    </Tooltip>
    <span
      v-if="fromToken && toToken"
      class="seperator"
    >
      /
    </span>
    <Tooltip
      v-if="toToken"
      class="symbol"
      :tooltip="toToken.name"
    >
      {{ shrinkString(toToken.symbol) }}
    </Tooltip>
  </span>
</template>

<script>
import Tooltip from './Tooltip.vue';
import AeIcon from '../../../icons/tokens/ae.svg';

export default {
  components: {
    Tooltip,
  },
  props: {
    /**
     * transactionTokenInfoResolvers []
     */
    tokens: { type: Array, required: true },
    symbolLength: { type: Number, default: 11 },
  },
  computed: {
    fromToken() {
      return this.tokens?.[0] ? this.mapToken(this.tokens[0]) : null;
    },
    toToken() {
      return this.tokens?.[1] ? this.mapToken(this.tokens[1]) : null;
    },
  },
  methods: {
    shrinkString(text) {
      return `${String(text).substring(0, this.symbolLength)}${text.length > this.symbolLength ? ' ...' : ''}`;
    },
    mapToken(token) {
      let img = `https://avatars.z52da5wt.xyz/${token.contractId}`;
      let imgBorder = true;

      if (token.isAe) {
        img = AeIcon;
        imgBorder = false;
      }

      return {
        ...token,
        symbol: token.isAe ? 'AE' : token.symbol,
        name: token.isAe ? 'Aeternity' : token.symbol,
        img,
        imgBorder,
      };
    },
  },
};
</script>

<style lang="scss" scoped>
@use '../../../styles/variables';
@use '../../../styles/typography';

.tokens {
  color: variables.$color-light-grey;
  display: inline-flex;
  align-items: center;

  @extend %face-sans-14-regular;

  .symbol {
    vertical-align: middle;
    white-space: nowrap;
  }

  .divider {
    color: variables.$color-blue;
    margin: 0 4px;
    word-break: break-all;
    vertical-align: middle;
  }

  .seperator {
    color: variables.$color-blue;
    margin: 0 1px;
    vertical-align: middle;
  }

  img {
    width: 16px;
    height: 16px;
    border-radius: 8px;
    margin-right: 6px;
    vertical-align: middle;

    &.border {
      border: 0.25px solid variables.$color-light-grey;
    }

    &:nth-child(2) {
      margin-left: -10px;
    }
  }
}
</style>
