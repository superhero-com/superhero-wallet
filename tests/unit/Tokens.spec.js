import { shallowMount } from '@vue/test-utils';
import Tokens from '../../src/popup/components/Tokens.vue';

let wrapper;

describe('Tokens', () => {
  beforeEach(() => {
    wrapper = shallowMount(Tokens, {
      propsData: {
        tokens: [{ symbol: 'test123', contractId: '1' }, { symbol: 'test456', contractId: '2' }],
        vertical: true,
      },
    });
  });

  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should test getAvailableCharLength', async () => {
    const getAvailableCharLength = jest.spyOn(wrapper.vm, 'getAvailableCharLength');
    let length = wrapper.vm.getAvailableCharLength();
    expect(getAvailableCharLength).toHaveBeenCalled();
    expect(length).toEqual(5);
    await wrapper.setProps({
      symbolLength: 11,
      doubleSymbolLength: 5,
      tokens: [{ symbol: 'test' }, { symbol: 'test456' }],
      vertical: true,
    });
    length = wrapper.vm.getAvailableCharLength();
    expect(length).toEqual(7);
  });
});
