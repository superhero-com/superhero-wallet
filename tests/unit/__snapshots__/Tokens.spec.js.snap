// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokens should match the snapshot 1`] = `
<span class="tokens rg vertical"><span class="icons"><img src="https://avatars.z52da5wt.xyz/2" title="test456" class="to-token border"> <img src="https://avatars.z52da5wt.xyz/1" title="test123" class="border pair"></span> <span class="symbols"><span class="symbol">
      test1...
    </span> <span class="separator">
      /
    </span> <span class="symbol">
      test4...
    </span></span></span>
`;
